pkg/stringid/stringid.go

```diff
@@ -80,16 +86,22 @@ func ValidateID(id string) error {
 
 func init() {
 	// safely set the seed globally so we generate random ids. Tries to use a
-	// crypto seed before falling back to time.
+	// crypto seed (concurrent-safe) before falling back to time (concurrent-unsafe).
 	var seed int64
 	if cryptoseed, err := cryptorand.Int(cryptorand.Reader, big.NewInt(math.MaxInt64)); err != nil {
 		// This should not happen, but worst-case fallback to time-based seed.
-		seed = time.Now().UnixNano()
+		// Attempt to mitigate clashes slightly by including the current thread ID.
+		// This is slow and imperfect, but preferable to hanging (indefinately)
+		// for the entropy pool to fill.
+		randLock.Lock()
+		seed = time.Now().UnixNano() + int64(syscall.Gettid())
```

Comment:
I don’t really see how this helps; the `UnixNano` seed has a 1e9 range, and this just slightly pertubes the value in that range but doesn’t change the range in any way. So, supposing the time window for two concurrent processes is, I dunno, 10 ms, and there is a 1/10e6 chance of a collision, AFAICS there should be almost exactly a 1/10e6 chance of collision after we add the TID, which is likely to be the same value plus or minus at most a few hundred. So how does this make a difference?

For this approach to a fix I’d like to see a much more detailed explanation for how this fails (if the theory is that there are two processes that _independently_ run this code and arrive at the same seed, how is such a prima facie extremely unlikely thing happening? Do we know how the `cryptoseed` initialization above fails, in the first place?), probably an explanation that somehow explains how the nanosecond values are actually multiples of some large number, and adding the smaller-granularity TID value actually adds a notable number of bits to the _entropy_ of the seed, decreasing the likelihood of failures.

Reply:
> how is such a prima facie extremely unlikely thing happening

And happening repeatedly, on multiple OSes and multiple versions :frowning_face: 

> Do we know how the cryptoseed initialization above fails

We do not, we're assuming kernel entropy is exhausted and my (bad) test-code analysis shows there is no other way for the same socket ID to clash across multiple tests.

Reply:
Oh...I should add:  The test VMs most certainly do **not** have access to the hardware RNG.  We're relying on the `rngd` service in Fedora, and nothing but luck in Ubuntu :sob: So it's easily possible just running the podman tests is soaking up all the entropy.

Reply:
I’m out of touch, but I thought that `/dev/urandom` reads don’t fail — they could block, but not fail. And the implementation linked at https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/crypto/rand/rand_unix.go doesn’t show any use of `O_NONBLOCK` anyway, I think.

So, some evidence to see that we are indeed taking this code path, and what the failure is, could be interesting.

Similarly, collecting a few hundred seed values (on the target platform) and seeing whether they happen to have suspiciously non-uniform value distribution, might be informative.

(Yeah, I could run these experiments myself, and I didn’t.)

---

Ultimately, understanding how exactly this happens would tickle my curiosity and interest, but I just don’t think it’s productive. If we add a duplicate check+retry to the caller, we can have pretty much any weak RNG here, even with a probability of collisions in the order of 1/1000, and it would work _reliably_. To me that’s a much more useful change than understanding in detail, whether the probability of collisions is 1/1e9, 1/1e6, 1/1e7, 2^-32, 2^-64 or something weird, and still having, in principle, the risk of occasional failures, however exactly we tune this.

Reply:
> doesn’t show any use of O_NONBLOCK anyway, I think.

It is also my understanding as well that `/dev/urandom` will block if there isn't enough entropy.  I haven't looked at the code at all (it scares me), I'm assuming somebody did when they added the error check and `time.nano()` fallback.

Since you know this code-base, would it make sense at all to issue some kind of warning (to the logs) when the fallback option is chosen?  That would at least increase observability of the actual underlying error.

> To me that’s a much more useful change than understanding in detail

Precisely, this PR isn't about making a better 20-sided die, it's about preventing process-to-process collisions.  If that is needed, and can be done centrally, it will benefit all callers of this function.